lib/srv/db/mysql/proxy.go

```diff
@@ -0,0 +1,160 @@
+	}
+	ctx, err = p.Middleware.WrapContextWithUser(ctx, tlsConn)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	serviceConn, err := p.Service.Connect(ctx, server.GetUser(), server.GetDatabase())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer serviceConn.Close()
+	// Before replying OK to the client which would make the client consider
+	// auth completed, wait for OK packet from db service indicating auth
+	// success.
+	err = p.waitForOK(server, serviceConn)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// Auth has completed, the client enters command phase, start proxying
+	// all messages back-and-forth.
+	err = p.Service.Proxy(ctx, tlsConn, serviceConn)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// credentialProvider is used by MySQL server created below.
+//
+// It's a no-op because authentication is done via mTLS.
+type credentialProvider struct{}
+
+func (p *credentialProvider) CheckUsername(_ string) (bool, error)         { return true, nil }
+func (p *credentialProvider) GetCredential(_ string) (string, bool, error) { return "", true, nil }
+
+// makeServer creates a MySQL server from the accepted client connection that
+// provides access to various parts of the handshake.
+func (p *Proxy) makeServer(clientConn net.Conn) *server.Conn {
+	return server.MakeConn(
+		clientConn,
+		server.NewServer(
+			serverVersion,
+			mysql.DEFAULT_COLLATION_ID,
+			mysql.AUTH_NATIVE_PASSWORD,
+			nil,
+			p.TLSConfig),
+		&credentialProvider{},
+		server.EmptyHandler{})
+}
+
+// performHandshake performs the initial handshake between MySQL client and
+// this server, up to the point where the client sends us a certificate for
+// authentication, and returns the upgraded connection.
+func (p *Proxy) performHandshake(server *server.Conn) (*tls.Conn, error) {
+	err := server.WriteInitialHandshake()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	err = server.ReadHandshakeResponse()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	// First part of the handshake completed and the connection has been
+	// upgraded to TLS so now we can look at the client certificate and
+	// see which database service to route the connection to.
+	tlsConn, ok := server.Conn.Conn.(*tls.Conn)
+	if !ok {
+		return nil, trace.BadParameter("expected tls connection")
+	}
+	return tlsConn, nil
+}
+
+// waitForOK waits for OK_PACKET from the database service which indicates
+// that auth on the other side completed successfully.
+func (p *Proxy) waitForOK(server *server.Conn, serviceConn net.Conn) error {
+	packet, err := protocol.ParsePacket(serviceConn)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	switch packet.(type) {
+	case *protocol.OK:
+		err = server.WriteOK(nil)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	case *protocol.Error:
+		err = server.WritePacket(packet.Bytes())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	default:
+		return trace.BadParameter("expected OK or ERR packet, got %s", packet)
```

Comment:
nit: do packets have helpful `fmt.Stringer` implementations?

Reply:
No, unfortunately. In the unlikely event this happens, it will dump the packet in its raw byte form so it will be possible to see which one we got.